pkg/controllers/policyrc/controller.go

```diff
@@ -20,175 +19,143 @@ package policyrc
+	overridePolicyInformer           fedcorev1a1informers.OverridePolicyInformer
+	clusterOverridePolicyInformer    fedcorev1a1informers.ClusterOverridePolicyInformer
+	propagationPolicyInformer        fedcorev1a1informers.PropagationPolicyInformer
+	clusterPropagationPolicyInformer fedcorev1a1informers.ClusterPropagationPolicyInformer
 
 	ppCounter, opCounter *Counter
 
 	// updates the local counter upon fed object updates
-	countWorker worker.ReconcileWorker
+	countWorker worker.ReconcileWorker[common.QualifiedName]
 	// pushes values from local counter to apiserver
-	persistPpWorker, persistOpWorker worker.ReconcileWorker
-
-	typeConfig *fedcorev1a1.FederatedTypeConfig
+	persistPpWorker, persistOpWorker worker.ReconcileWorker[common.QualifiedName]
 
+	client  generic.Client
 	metrics stats.Metrics
 	logger  klog.Logger
 }
 
-func StartController(controllerConfig *util.ControllerConfig,
-	stopChan <-chan struct{}, typeConfig *fedcorev1a1.FederatedTypeConfig,
-) error {
-	controller, err := newController(controllerConfig, typeConfig)
-	if err != nil {
-		return err
-	}
-	controller.logger.Info("Starting policyrc controller")
-	controller.Run(stopChan)
-	return nil
-}
-
-func newController(controllerConfig *util.ControllerConfig,
-	typeConfig *fedcorev1a1.FederatedTypeConfig,
+func NewPolicyRCController(
+	restConfig *rest.Config,
+	fedInformerFactory fedinformers.SharedInformerFactory,
+	metrics stats.Metrics,
+	logger klog.Logger,
+	workerCount int,
 ) (*Controller, error) {
-	federatedAPIResource := typeConfig.GetFederatedType()
-
-	userAgent := fmt.Sprintf("%s-policyrc-controller", strings.ToLower(federatedAPIResource.Kind))
-	configWithUserAgent := rest.CopyConfig(controllerConfig.KubeConfig)
-	rest.AddUserAgent(configWithUserAgent, userAgent)
 
 	c := &Controller{
-		name:       userAgent,
-		typeConfig: typeConfig,
-		metrics:    controllerConfig.Metrics,
-		logger:     klog.LoggerWithValues(klog.Background(), "controller", ControllerName, "ftc", typeConfig.Name),
+		client:                           generic.NewForConfigOrDie(restConfig),
+		fedObjectInformer:                fedInformerFactory.Core().V1alpha1().FederatedObjects(),
+		clusterFedObjectInformer:         fedInformerFactory.Core().V1alpha1().ClusterFederatedObjects(),
+		propagationPolicyInformer:        fedInformerFactory.Core().V1alpha1().PropagationPolicies(),
+		clusterPropagationPolicyInformer: fedInformerFactory.Core().V1alpha1().ClusterPropagationPolicies(),
+		overridePolicyInformer:           fedInformerFactory.Core().V1alpha1().OverridePolicies(),
+		clusterOverridePolicyInformer:    fedInformerFactory.Core().V1alpha1().ClusterOverridePolicies(),
+		metrics:                          metrics,
+		logger:                           logger.WithValues("controller", ControllerName),
 	}
 
-	c.countWorker = worker.NewReconcileWorker(
+	c.fedObjectInformer.Informer().AddEventHandler(util.NewTriggerOnAllChanges(func(o pkgruntime.Object) {
```

Comment:
Let's make it consistent whether we want to handle the error from `AddEventHandler` or not. The error is handled below.

Reply:
Let's also move this below with the rest of the `AddEventHandler` calls.